// @aerisweather/javascript-sdk/dist/geo/CoordinateBounds.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.normalize = void 0;

var _InvalidCoordinateError = _interopRequireDefault(require("../errors/InvalidCoordinateError"));

var _utils = require("../utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var normalize = function (coord) {
  var factor = Math.ceil(Math.abs(coord.lat) / 180);
  if (coord.lat > 90) coord.lat -= 180 * factor;else if (coord.lat < -90) coord.lat += 180 * factor;
  factor = Math.ceil(Math.abs(coord.lon) / 360);
  if (coord.lon > 180) coord.lon -= 360 * factor;else if (coord.lon < -180) coord.lon += 360 * factor;
  return coord;
};
/**
 * A `CoordinateBounds` object represents a rectangular region defined by northwest and southeast
 * coordinates.
 */


exports.normalize = normalize;

var CoordinateBounds =
/** @class */
function () {
  /**
   * Initializes a bounds instance defined by the specified northwest and southeast coordinates.
   * @param nw
   * @param se
   */
  function CoordinateBounds(nw, se) {
    /** The northern latitude */
    this.north = 0;
    /** The southern latitude */

    this.south = 0;
    /** The western longitude */

    this.west = 0;
    /** The eastern longitude */

    this.east = 0;
    nw = normalize(nw);
    se = normalize(se);

    if (undefined === nw || !(0, _utils.isObject)(nw)) {
      throw new _InvalidCoordinateError.default('CoordinateBounds - northwest coordinate value is required');
    }

    if (undefined === se || !(0, _utils.isObject)(se)) {
      throw new _InvalidCoordinateError.default('CoordinateBounds - southeast coordinate value is required');
    }

    if (nw.lat < se.lat) {
      throw new _InvalidCoordinateError.default('CoordinateBounds - northern latitude must be greater than southern latitude');
    } // if (nw.lon > se.lon) {
    // 	throw new InvalidCoordinateError('CoordinateBounds - eastern longitude must be greater than western longitude');
    // }


    this.north = nw.lat;
    this.south = se.lat;
    this.west = nw.lon;
    this.east = se.lon;
  }

  CoordinateBounds.fromBounds = function (bounds) {
    return new CoordinateBounds({
      lat: bounds.north,
      lon: bounds.west
    }, {
      lat: bounds.south,
      lon: bounds.east
    });
  };

  CoordinateBounds.fromPoints = function (points) {
    var bounds = {
      north: null,
      south: null,
      west: null,
      east: null
    };
    points.forEach(function (coord, index) {
      if (index === 0) {
        bounds.north = coord.lat;
        bounds.south = coord.lat;
        bounds.west = coord.lon;
        bounds.east = coord.lon;
      } else if ((0, _utils.isset)(coord.lat) && (0, _utils.isset)(coord.lon)) {
        bounds.north = Math.max(bounds.north, coord.lat);
        bounds.south = Math.min(bounds.south, coord.lat);
        bounds.west = Math.min(bounds.west, coord.lon);
        bounds.east = Math.max(bounds.east, coord.lon);
      }
    });
    return CoordinateBounds.fromBounds(bounds);
  };
  /**
   * Returns the northwestern coordinate for the bounds.
   */


  CoordinateBounds.prototype.northwest = function () {
    return {
      lat: this.north,
      lon: this.west
    };
  };
  /**
   * Returns the southwestern coordinate for the bounds.
   */


  CoordinateBounds.prototype.southwest = function () {
    return {
      lat: this.south,
      lon: this.west
    };
  };
  /**
   * Returns the northeastern coordinate for the bounds.
   */


  CoordinateBounds.prototype.northeast = function () {
    return {
      lat: this.north,
      lon: this.east
    };
  };
  /**
   * Returns the southeastern coordinate for the bounds.
   */


  CoordinateBounds.prototype.southeast = function () {
    return {
      lat: this.south,
      lon: this.east
    };
  };
  /**
   * Returns the center coordinate for the bounds.
   */


  CoordinateBounds.prototype.center = function () {
    return {
      lat: this.north - (this.north - this.south) / 2,
      lon: this.west - (this.west - this.east) / 2
    };
  };
  /**
   * Expands the bounds to include the specified coordinate. If the bounds already contains the
   * coodinate, then the bounds will not change.
   *
   * @param coord
   */


  CoordinateBounds.prototype.extend = function (coord) {
    if (!this.north && !this.south && !this.west && !this.east) {
      this.north = coord.lat;
      this.south = coord.lat;
      this.west = coord.lon;
      this.east = coord.lon;
      return;
    }

    this.north = Math.max(this.north, coord.lat);
    this.south = Math.min(this.south, coord.lat);
    this.west = Math.min(this.west, coord.lon);
    this.east = Math.max(this.east, coord.lon);
  };
  /**
   * Returns whether the bounds intersects the specified bounds but does not contain it.
   * @param bounds
   */


  CoordinateBounds.prototype.intersects = function (bounds) {
    if (this.contains(bounds)) {
      return false;
    }

    var north = bounds.north,
        south = bounds.south,
        east = bounds.east,
        west = bounds.west;
    return (north <= this.north && north >= this.south || south >= this.south && south <= this.north) && ( // eslint-disable-line max-len
    west >= this.west && west <= this.east || east <= this.east && east >= this.west);
  };
  /**
   * Returns whether the bounds contains the specified coordinate or bounds.
   * @param value
   */


  CoordinateBounds.prototype.contains = function (value) {
    if (value instanceof CoordinateBounds) {
      var _a = value,
          north = _a.north,
          south = _a.south,
          east = _a.east,
          west = _a.west;
      return north < this.north && south > this.south && west > this.west && east < this.east;
    }

    var _b = value,
        lat = _b.lat,
        lon = _b.lon;
    return lat < this.north && lat > this.south && lon > this.west && lon < this.east;
  };
  /**
   * Returns a string representation of the coordinate bounds.
   */


  CoordinateBounds.prototype.toString = function () {
    return this.north + "," + this.west + "," + this.south + "," + this.east;
  };

  return CoordinateBounds;
}();

var _default = CoordinateBounds;
exports.default = _default;
