// @aerisweather/javascript-sdk/dist/maps/interactive/strategies/google/layers/Layer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _EventDispatcher = _interopRequireDefault(require("../../../../../event/EventDispatcher"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var __extends = void 0 && (void 0).__extends || function () {
  var extendStatics = function (d, b) {
    extendStatics = Object.setPrototypeOf || {
      __proto__: []
    } instanceof Array && function (d, b) {
      d.__proto__ = b;
    } || function (d, b) {
      for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    };

    return extendStatics(d, b);
  };

  return function (d, b) {
    extendStatics(d, b);

    function __() {
      this.constructor = d;
    }

    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
  };
}();
/* eslint-disable spaced-comment */
/// <reference path="../../../../../../node_modules/@types/googlemaps/index.d.ts" />

/* eslint-enable spaced-comment */


/**
 * `Layer` is a container class that is responsible for creating and managing a Google Maps
 * source and associated layer representation on a Google Maps map.
 */
var GoogleLayer =
/** @class */
function (_super) {
  __extends(GoogleLayer, _super);

  function GoogleLayer(layer, id) {
    if (id === void 0) {
      id = undefined;
    }

    var _this = _super.call(this) || this;

    _this._hidden = false;
    _this.id = "aeris-layer-" + layer + (id ? "-" + id : '');
    return _this;
  }

  Object.defineProperty(GoogleLayer.prototype, "map", {
    /**
     * Map currently associated with the layer.
     *
     * @readonly
     * @type {google.maps.Map}
     * @memberof Layer
     */
    get: function () {
      return this._map;
    },
    enumerable: true,
    configurable: true
  });
  Object.defineProperty(GoogleLayer.prototype, "isHidden", {
    /**
     * Returns whether the layer is currently hidden.
     *
     * @readonly
     * @type {boolean}
     * @memberof Layer
     */
    get: function () {
      return this._hidden;
    },
    enumerable: true,
    configurable: true
  });
  /**
   * Adds the layer to the specified map.
   *
   * @param {google.maps.Map} map
   * @param {number} index
   * @memberof Layer
   */

  GoogleLayer.prototype.addTo = function (map, index) {
    if (this.map === map) return;
    this._map = map;
    this._targetIndex = index;
  };
  /**
   * Removes the layer from the map it is currently active on.
   *
   * @returns
   * @memberof Layer
   */


  GoogleLayer.prototype.remove = function () {
    if (!this.map) return;
    this._map = null;
  };
  /**
   * Shows the layer on the map.
   *
   * @memberof Layer
   */


  GoogleLayer.prototype.show = function () {
    this._hidden = false;
  };
  /**
   * Hides the layer on the map.
   *
   * @memberof Layer
   */


  GoogleLayer.prototype.hide = function () {
    this._hidden = true;
  };
  /**
   * Changes the opacity of the layer.
   *
   * @param {number} value
   * @memberof Layer
   */


  GoogleLayer.prototype.setOpacity = function (value) {
    throw new Error('Subclasses of GoogleLayer should override setOpacity()');
  };

  GoogleLayer.prototype.setZIndex = function (index) {
    throw new Error('Subclasses of GoogleLayer should override setZIndex()');
  };
  /**
   * Returns whether the layer currently exists on the map.
   *
   * @returns {boolean}
   * @memberof Layer
   */


  GoogleLayer.prototype.exists = function () {
    var map = this.map;

    if (!map) {
      return false;
    } // return (map.getLayer(this.layerId) !== undefined);


    return false;
  };
  /**
   * Forces a redraw of the layer on the map.
   *
   * @returns
   * @memberof Layer
   */


  GoogleLayer.prototype.invalidate = function () {
    throw new Error('Subclasses must override `invalidate`');
  };

  return GoogleLayer;
}(_EventDispatcher.default);

var _default = GoogleLayer;
exports.default = _default;
module.exports = exports.default;
