// @aerisweather/javascript-sdk/dist/utils/color.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isLight = exports.luminance = exports.toRGB = void 0;

/**
 * Converts the specified hex color string to an RGB object.
 *
 * @param {string} hex
 * @returns {RGB}
 */
var toRGB = function (hex) {
  hex = hex.replace(/^#/, '');
  var color = {
    r: 0,
    g: 0,
    b: 0
  }; // eslint-disable-line object-curly-newline

  if (hex.length === 6) {
    color.r = parseInt(hex.substring(0, 2), 16);
    color.g = parseInt(hex.substring(2, 4), 16);
    color.b = parseInt(hex.substring(4), 16);
  }

  return color;
};
/**
 * Returns the luminance of the specified color.
 *
 * The value for `color` can either be a hex color string, `#ffcc00`, or an RGB object,
 * `{ r: 100, g: 150, b: 50 }`.
 *
 * @param {string | RGB} color
 * @returns {number}
 */


exports.toRGB = toRGB;

var luminance = function (color) {
  var rgb = toRGB(color);
  if (!rgb) return null;
  return 0.2126 * rgb.r + 0.7152 * rgb.g + 0.0722 * rgb.b;
};
/**
 * Returns whether the specified color is light based on its luminance.
 *
 * The value for `color` can either be a hex color string, `#ffcc00`, or an RGB object,
 * `{ r: 100, g: 150, b: 50 }`.
 *
 * @param {string | RGB} color
 * @returns {boolean}
 */


exports.luminance = luminance;

var isLight = function (color) {
  var lum = luminance(color);
  return lum && lum >= 140;
};

exports.isLight = isLight;
