// @aerisweather/javascript-sdk/dist/utils/layout.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.viewportSizeClass = exports.getPinnedPosition = void 0;

var _ = require(".");

var getPinnedPosition = function (position, size, container, translate) {
  var x = 0;
  var y = 0;
  translate = (0, _.extend)({
    x: 0,
    y: 0
  }, translate);

  if ((0, _.isString)(position)) {
    var anchor = position;
    x = (container.width - size.width) / 2;
    y = (container.height - size.height) / 2;

    if (/^top/.test(anchor)) {
      y = 0;
    } else if (/^bottom/.test(anchor)) {
      y = container.height - size.height;
    }

    if (/left$/.test(anchor)) {
      x = 0;
    } else if (/right$/.test(anchor)) {
      x = container.width - size.width;
    }
  } else {
    var p = position;
    x = p.x;
    y = p.y;
  }

  x += translate.x;
  y += translate.y;
  return {
    x: x,
    y: y
  };
};

exports.getPinnedPosition = getPinnedPosition;

var viewportSizeClass = function () {
  if (!window) return {
    width: null,
    height: null
  };
  var width = 'normal';
  var height = 'normal';

  if (window.matchMedia("(max-width: 500px)").matches) {
    width = 'compact';
  } else if (window.matchMedia("(min-width: 1000px)").matches) {
    width = 'desktop';
  }

  if (window.matchMedia("(max-height: 500px)").matches) {
    height = 'compact';
  } else if (window.matchMedia("(min-height: 1000px)").matches) {
    height = 'desktop';
  }

  return {
    width: width,
    height: height
  };
};

exports.viewportSizeClass = viewportSizeClass;
