// @aerisweather/javascript-sdk/dist/maps/interactive/style/layers/airquality.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.config = void 0;

var _utils = require("../../../../utils");

/**
 * @private
*/
var color = function (code) {
  code = code.toLowerCase().replace(/\s/, '-');

  switch (code) {
    case 'good':
      return '#29e11f';

    case 'moderate':
      return '#f8f92a';

    case 'usg':
      return '#f9681b';

    case 'unhealthy':
      return '#f60115';

    case 'very-unhealthy':
      return '#7a2c83';

    case 'hazardous':
      return '#65001b';

    default:
      return '#000000';
  }
};
/**
 * @private
*/


var size = function (code) {
  code = code.toLowerCase().replace(/\s/, '-');

  switch (code) {
    case 'moderate':
      return 12;

    case 'usg':
      return 14;

    case 'unhealthy':
      return 18;

    case 'very-unhealthy':
      return 20;

    case 'hazardous':
      return 24;

    default:
      return 10;
  }
};
/**
 * @private
*/


var config = {
  request: {
    endpoint: 'airquality'
  },
  style: {
    marker: function (data) {
      var period = (0, _utils.get)(data, 'periods.0');

      if (!period) {
        return null;
      }

      var type = (0, _utils.get)(period, 'category');

      if (!type) {
        return null;
      }

      var width = size(type);
      return {
        className: 'marker-airquality',
        svg: {
          shape: {
            type: 'circle',
            fill: {
              color: color(type)
            },
            stroke: {
              color: '#ffffff',
              width: 2
            }
          }
        },
        size: [width, width]
      };
    }
  },
  data: {
    properties: {
      id: 'id',
      timestamp: 'timestamp'
    }
  }
};
exports.config = config;
var _default = config;
exports.default = _default;
