// @aerisweather/javascript-sdk/dist/maps/interactive/style/layers/fireOutlook.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.config = void 0;

var _ApiRequest = require("../../../../network/api/ApiRequest");

var _utils = require("../../../../utils");

var color = function (code) {
  code = code.toLowerCase();

  switch (code) {
    case 'elevated':
      return '#ffb367';

    case 'critical':
      return '#ff666a';

    case 'extreme':
      return '#fe59ff';

    default:
      return '#000000';
  }
};
/**
 * @private
*/


var config = {
  request: {
    endpoint: 'fires/outlook',
    action: _ApiRequest.ApiAction.SEARCH,
    parameters: {
      filter: 'geo'
    }
  },
  style: {
    polygon: function (data) {
      var type = (0, _utils.get)(data, 'details.risk.type');
      return {
        fill: {
          color: color(type),
          opacity: 0.7
        }
      };
    }
  },
  data: {
    properties: {
      id: 'id',
      path: 'geoPoly',
      category: 'details.risk.type',
      timestamp: {
        from: 'details.range.minTimestamp',
        to: 'details.range.maxTimestamp'
      }
    }
  }
};
exports.config = config;
var _default = config;
exports.default = _default;
