// @aerisweather/javascript-sdk/dist/maps/interactive/style/layers/lightningStrikes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.config = void 0;

var _utils = require("../../../../utils");

/**
 * @private
*/
var color = function (delta) {
  if (delta < 1) {
    return '#f5f5f5';
  }

  if (delta < 5) {
    return '#dddddd';
  }

  if (delta < 10) {
    return '#bbbbbb';
  }

  if (delta < 15) {
    return '#888888';
  }

  return '#545454';
};
/**
 * @private
*/


var config = {
  request: {
    endpoint: 'lightning'
  },
  style: {
    marker: function (data) {
      var ts = (0, _utils.get)(data, 'ob.timestamp');

      if (!ts) {
        return null;
      }

      var delta = (new Date().getTime() - ts * 1000) / (1000 * 60);
      return {
        className: 'marker-record',
        svg: {
          shape: {
            type: 'circle',
            fill: {
              color: color(delta)
            },
            stroke: {
              color: '#ffffff',
              width: 2
            }
          }
        },
        size: [14, 14]
      };
    }
  },
  data: {
    properties: {
      id: 'id',
      timestamp: 'ob.timestamp'
    }
  }
};
exports.config = config;
var _default = config;
exports.default = _default;
