// @aerisweather/javascript-sdk/dist/maps/interactive/style/layers/stormcells.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.config = void 0;

var _utils = require("../../../../utils");

var color = function (code) {
  code = code.toLowerCase();

  switch (code) {
    case 'general':
      return '#2ED300';

    case 'hail':
      return '#EBE100';

    case 'rotating':
      return '#F17200';

    case 'tornado':
      return '#FF2600';

    default:
      return '#000000';
  }
};
/**
 * @private
*/


var config = {
  request: {
    endpoint: 'stormcells'
  },
  style: {
    marker: function (data) {
      var type = (0, _utils.get)(data, 'traits.type');
      return {
        className: 'marker-stormcell',
        svg: {
          shape: {
            type: 'circle',
            fill: {
              color: color(type)
            },
            stroke: {
              color: '#ffffff',
              width: 2
            }
          }
        },
        size: [12, 12]
      };
    },
    polygon: function (data) {
      var type = (0, _utils.get)(data, 'traits.type');
      return {
        fill: {
          color: color(type),
          opacity: 0.5
        }
      };
    }
  },
  data: {
    properties: {
      id: 'id',
      path: 'forecast.cone.wide',
      category: 'traits.type',
      timestamp: 'ob.timestamp'
    },
    reversedCoord: true
  }
};
exports.config = config;
var _default = config;
exports.default = _default;
