// @aerisweather/javascript-sdk/dist/maps/interactive/MapLayerConfig.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var defaultStyles = _interopRequireWildcard(require("./style"));

var _Layers = require("../Layers");

var _utils = require("../../utils");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

/**
 * Class that manages the configuration and style providers based on layer code.
 *
 * @export
 * @class MapLayerConfig
 */
var MapLayerConfig =
/** @class */
function () {
  /**
   * Initializes and returns a map configuration instance with the provided styles.
   *
   * @param {({ [key: string]: LayerStyleGroups | LayerStyle })} [styles]
   * @memberof MapLayerConfig
   */
  function MapLayerConfig(styles) {
    this._configs = {
      'air-quality': defaultStyles.airquality,
      'convective': defaultStyles.convectiveOutlook,
      'drought-monitor': defaultStyles.droughtMonitor,
      earthquakes: defaultStyles.earthquakes,
      fires: defaultStyles.fires,
      'fire-outlook': defaultStyles.fireOutlook,
      'lightning-strikes': defaultStyles.lightningStrikes,
      records: defaultStyles.records,
      'river-observations': defaultStyles.rivers,
      stormcells: defaultStyles.stormcells,
      stormreports: defaultStyles.stormreports,
      text: defaultStyles.text,
      'tropical-cyclones': defaultStyles.tropicalcyclones,
      'tropical-cyclones-break-points': defaultStyles.tropicalBreakpoints
    };
    if (styles) this.setStyles(styles);
  }
  /**
   * Returns the layer configuration for the specified layer code.
   *
   * @param {string} layer
   * @returns {LayerOptions}
   * @memberof MapLayerConfig
   */


  MapLayerConfig.prototype.getConfig = function (layer) {
    return this._configs[layer];
  };

  MapLayerConfig.prototype.setConfig = function (layer, config) {
    this._configs[layer] = config;
  };
  /**
   * Sets the style(s) for the specified layer code.
   *
   * @param {string} layer
   * @param {(LayerStyleGroups | LayerStyle)} style
   * @memberof MapLayerConfig
   */


  MapLayerConfig.prototype.setStyle = function (layer, style) {
    var config = this.getConfig(layer) || {};

    if ((0, _Layers.isTile)(layer)) {
      var rasterStyle = void 0;

      if (style.raster) {
        rasterStyle = style.raster;
      } else {
        rasterStyle = style;
      }

      config.style = (0, _utils.extend)(config.style, rasterStyle);
    } else if ((0, _Layers.isVector)(layer) || (0, _Layers.isText)(layer)) {
      var markerStyle = void 0;
      var polygonStyle = void 0;
      var polylineStyle = void 0;
      var grouped = (0, _utils.isset)(style.marker) || (0, _utils.isset)(style.polygon) || (0, _utils.isset)(style.polyline);

      if (style.marker) {
        markerStyle = style.marker;
      } else if (grouped === false) {
        markerStyle = style;
      }

      if (style.polygon) {
        polygonStyle = style.polygon;
      } else if (grouped === false) {
        polygonStyle = style;
      }

      if (style.polyline) {
        polylineStyle = style.polyline;
      } else if (grouped === false) {
        polylineStyle = style;
      }

      config.style = (0, _utils.extend)(config.style, {
        marker: markerStyle,
        polygon: polygonStyle,
        polyline: polylineStyle
      });
    }

    this._configs[layer] = (0, _utils.cloneDeep)(config);
  };
  /**
   * Sets the style(s) for the specified layers.
   *
   * @param {({ [layer: string]: LayerStyleGroups | LayerStyle })} layers
   * @memberof MapLayerConfig
   */


  MapLayerConfig.prototype.setStyles = function (layers) {
    var _this = this;

    Object.keys(layers).forEach(function (layer) {
      return _this.setStyle(layer, layers[layer]);
    });
  };
  /**
   * Returns the style associated with the specified layer code, if any.
   *
   * @param {string} layer
   * @param {(( 'raster' | 'marker' | 'polygon' | 'polyline'))} type
   * @returns {LayerStyle}
   * @memberof MapLayerConfig
   */


  MapLayerConfig.prototype.getStyle = function (layer, type) {
    var config = this.getConfig(layer);

    if (config) {
      var style = config.style;

      if (style[type]) {
        return style[type];
      }

      return style;
    }

    return undefined;
  };
  /**
   * Sets the text style to use for text annotations.
   *
   * @param {ShapeStyle} style
   * @memberof MapLayerConfig
   */


  MapLayerConfig.prototype.setTextStyle = function (style) {
    this.setStyle('text', {
      marker: style
    });
  };
  /**
   * Returns the text style.
   *
   * @returns {ShapeStyle}
   * @memberof MapLayerConfig
   */


  MapLayerConfig.prototype.getTextStyle = function () {
    return this.getStyle('text', 'marker');
  };

  return MapLayerConfig;
}();

var _default = MapLayerConfig;
exports.default = _default;
module.exports = exports.default;
