// @aerisweather/javascript-sdk/dist/account/Account.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _Permissions = _interopRequireDefault(require("./Permissions"));

var _ApiRequest = _interopRequireDefault(require("../network/api/ApiRequest"));

var _MapRequest = _interopRequireDefault(require("../network/maps/MapRequest"));

var _globals = require("../globals");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// let _cache: {
// 	[key: string]: Account
// } = {};
var Account =
/** @class */
function () {
  function Account(id, secret) {
    this.id = id;
    this.secret = secret;
    this.servers = _globals.servers;
    this._permissions = new _Permissions.default(id, secret);
  }

  Account.prototype.credentials = function () {
    return {
      id: this.id,
      secret: this.secret
    };
  };

  Account.prototype.permissions = function () {
    return this._permissions;
  };

  Account.prototype.canAccess = function (endpoint) {
    return this.permissions().canAccess(endpoint);
  };
  /**
   * Creates and returns a new ApiRequest configured for the account.
   *
   * @returns {ApiRequest}
   * @memberof Account
   */


  Account.prototype.api = function () {
    return new _ApiRequest.default({
      server: this.servers.api,
      client: {
        id: this.id,
        secret: this.secret
      }
    });
  };
  /**
   * Creates and returns a new MapRequest configured for the account.
   *
   * @returns {MapRequest}
   * @memberof Account
   */


  Account.prototype.map = function () {
    return new _MapRequest.default({
      server: this.servers.maps,
      client: {
        id: this.id,
        secret: this.secret
      }
    });
  };

  return Account;
}();

var _default = Account;
exports.default = _default;
module.exports = exports.default;
