// @aerisweather/javascript-sdk/dist/apps/views/TimelinePanel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _dateFns = require("date-fns");

var _Panel = _interopRequireDefault(require("../ui/Panel"));

var _TimeSlider = _interopRequireDefault(require("./TimeSlider"));

var _ActivityIndicator = _interopRequireDefault(require("../../ui/ActivityIndicator"));

var _RangeSlider = _interopRequireDefault(require("../ui/RangeSlider"));

var _Tooltip = _interopRequireDefault(require("../ui/Tooltip"));

var _utils = require("../../utils");

var _layout = require("../../utils/layout");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var __extends = void 0 && (void 0).__extends || function () {
  var extendStatics = function (d, b) {
    extendStatics = Object.setPrototypeOf || {
      __proto__: []
    } instanceof Array && function (d, b) {
      d.__proto__ = b;
    } || function (d, b) {
      for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    };

    return extendStatics(d, b);
  };

  return function (d, b) {
    extendStatics(d, b);

    function __() {
      this.constructor = d;
    }

    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
  };
}();

/**
 * A `TimelinePanel` is a panel component that renders and manages a timeline animation slider
 * control that allows a user to scrub an animation's current time position and change the
 * animation start/end time range.
 *
 * @export
 * @class TimelinePanel
 * @extends {Panel<TimelinePanelOptions, TimelinePanelState>}
 */
var TimelinePanel =
/** @class */
function (_super) {
  __extends(TimelinePanel, _super);
  /**
   * Initializes and returns a timeline panel instance configured with the specified animation and options.
   * @param {Animation} animation
   * @param {TimelinePanelOptions} [opts]
   * @memberof TimelinePanel
   */


  function TimelinePanel(animation, opts) {
    var _this = this;

    var customMarks = (0, _utils.get)(opts, 'range.marks');
    opts = (0, _utils.extend)({
      range: {},
      ui: {
        controls: '.awxjs__app__ui-panel-timeline-center',
        timeline: '.awxjs__app__ui-panel-timeline-right',
        play: '.awxjs__ui-btn-play',
        now: '.awxjs__ui-btn-now',
        time: '.awxjs__app__ui-panel-timeline-time',
        settings: '.awxjs__ui-btn-settings'
      }
    }, opts, {
      className: opts && opts.className ? "app__ui-panel-timeline " + opts.className : 'app__ui-panel-timeline'
    });
    _this = _super.call(this, opts) || this;
    var vp = (0, _layout.viewportSizeClass)();
    _this._animation = animation;
    _this._timeline = new _TimeSlider.default(animation);
    _this._indicator = new _ActivityIndicator.default();
    _this._slider = new _RangeSlider.default((0, _utils.extend)({
      title: 'Time Range <span>(hours)</span>',
      width: vp.width === 'compact' ? 200 : 300,
      value: [-6, 0],
      range: {
        min: -24,
        max: 24
      },
      marksEvery: (0, _utils.get)(opts, 'range.marksEvery') || '% 12',
      marks: customMarks || {
        0: 'Now'
      },
      step: vp.width === 'compact' ? 6 : 3,
      tooltip: function (value) {
        if (value === 0) {
          return 'Now';
        }

        var prefix = value > 0 ? '+' : '';
        return "" + prefix + value + " hours";
      }
    }, _this.opts.range));
    return _this;
  }

  TimelinePanel.prototype._render = function () {
    return "\n\t\t\t<div class=\"awxjs__app__ui-panel-container\">\n\t\t\t\t<div class=\"awxjs__app__ui-panel-content\">\n\t\t\t\t\t<div class=\"awxjs__ui-cols align-center\">\n\t\t\t\t\t\t<div class=\"awxjs__app__ui-panel-timeline-left\">\n\t\t\t\t\t\t\t<div class=\"awxjs__app__ui-panel-timeline-time\"></div>\n\t\t\t\t\t\t</div>\n\t\t\t\t\t\t<div class=\"awxjs__app__ui-panel-timeline-center\">\n\t\t\t\t\t\t\t<div class=\"awxjs__app__ui-panel-timeline-controls awxjs__ui-row align-center\">\n\t\t\t\t\t\t\t\t<button class=\"awxjs__ui-btn awxjs__ui-btn-icon awxjs__ui-btn-play\">\n\t\t\t\t\t\t\t\t\t<svg width=\"100pt\" height=\"100pt\" version=\"1.1\" viewBox=\"0 0 100 100\" xmlns=\"http://www.w3.org/2000/svg\">\n\t\t\t\t\t\t\t\t\t\t<path transform=\"scale(.3003)\" d=\"m260 166.99l-93.006 54.008-93.994 52.994v-214.99l93.994 53.995zm0 0\" fill-rule=\"evenodd\" />\n\t\t\t\t\t\t\t\t\t</svg>\n\t\t\t\t\t\t\t\t</button>\n\t\t\t\t\t\t\t\t<button class=\"awxjs__ui-btn awxjs__ui-btn-now\">Now</button>\n\t\t\t\t\t\t\t</div>\n\t\t\t\t\t\t</div>\n\t\t\t\t\t\t<div class=\"awxjs__app__ui-panel-timeline-right\"></div>\n\t\t\t\t\t</div>\n\t\t\t\t\t<div class=\"awxjs__ui-btn-icon awxjs__ui-btn-settings\">\n\t\t\t\t\t\t<svg version=\"1.1\" viewBox=\"0 0 100 100\" xmlns=\"http://www.w3.org/2000/svg\">\n\t\t\t\t\t\t\t<path d=\"m94.801 40.801-10.602-1.8008c-0.60156-1.8984-1.3984-3.6992-2.3008-5.5l6.1992-8.6992c0.89844-1.3008 0.80078-3.1016-0.30078-4.1992l-8.5-8.5c-1.1016-1.1016-2.8984-1.3008-4.1992-0.30078l-8.6992 6.1992c-1.6992-0.89844-3.6016-1.6992-5.5-2.3008l-1.6992-10.5c-0.30078-1.6016-1.6016-2.6992-3.1992-2.6992h-12c-1.6016 0-2.8984 1.1016-3.1992 2.6992l-1.8008 10.602c-1.8984 0.60156-3.6992 1.3984-5.5 2.3008l-8.6992-6.1992c-1.3008-0.89844-3.1016-0.80078-4.1992 0.30078l-8.5 8.5c-1.1016 1.1016-1.3008 2.8984-0.30078 4.1992l6.1992 8.6992c-0.89844 1.6992-1.6992 3.6016-2.3008 5.5l-10.5 1.6992c-1.6016 0.30078-2.6992 1.6016-2.6992 3.1992v12c0 1.6016 1.1016 2.8984 2.6992 3.1992l10.602 1.8008c0.60156 1.8984 1.3984 3.6992 2.3008 5.5l-6.1992 8.6992c-0.89844 1.3008-0.80078 3.1016 0.30078 4.1992l8.5 8.5c1.1016 1.1016 2.8984 1.3008 4.1992 0.30078l8.6992-6.1992c1.6992 0.89844 3.6016 1.6992 5.5 2.3008l1.8008 10.602c0.30078 1.6016 1.6016 2.6992 3.1992 2.6992h12c1.6016 0 2.8984-1.1016 3.1992-2.6992l1.6992-10.703c1.8984-0.60156 3.6992-1.3984 5.5-2.3008l8.6992 6.1992c1.3008 0.89844 3.1016 0.80078 4.1992-0.30078l8.5-8.5c1.1016-1.1016 1.3008-2.8984 0.30078-4.1992l-6.1992-8.6992c0.89844-1.6992 1.6992-3.6016 2.3008-5.5l10.602-1.8008c1.6016-0.30078 2.6992-1.6016 2.6992-3.1992v-11.898c-0.10156-1.6016-1.2031-2.8984-2.8008-3.1992zm-44.801 26.301c-9.5 0-17.102-7.6992-17.102-17.102 0-9.5 7.6992-17.102 17.102-17.102 9.3984 0 17.102 7.6016 17.102 17.102s-7.6016 17.102-17.102 17.102z\"/>\n\t\t\t\t\t\t</svg>\n\t\t\t\t\t</div>\n\t\t\t\t</div>\n\t\t\t</div>\n\t\t";
  };

  TimelinePanel.prototype._onAddedToDOM = function () {
    this._timeline.addTo(this.ui.timeline);

    this._indicator.appendTo(this.ui.timeline);

    this._setTime(this._animation.currentTime);
  };

  TimelinePanel.prototype._setupEvents = function () {
    var _this = this;

    this._animation.on('load:start play', function () {
      _this.ui.play.html("\n\t\t\t\t<svg width=\"100pt\" height=\"100pt\" version=\"1.1\" viewBox=\"0 0 100 100\" xmlns=\"http://www.w3.org/2000/svg\">\n\t\t\t\t\t<path d=\"m76.5 15h-53c-4.6875 0-8.5 3.8125-8.5 8.5v53c0 4.6875 3.8125 8.5 8.5 8.5h53c4.6875 0 8.5-3.8125 8.5-8.5v-53c0-4.6875-3.8125-8.5-8.5-8.5z\"/>\n\t\t\t\t</svg>\n\t\t\t");
    }).on('stop', function () {
      _this.ui.play.html("\n\t\t\t\t<svg width=\"100pt\" height=\"100pt\" version=\"1.1\" viewBox=\"0 0 100 100\" xmlns=\"http://www.w3.org/2000/svg\">\n\t\t\t\t\t<path transform=\"scale(.3003)\" d=\"m260 166.99l-93.006 54.008-93.994 52.994v-214.99l93.994 53.995zm0 0\" fill-rule=\"evenodd\" />\n\t\t\t\t</svg>\n\t\t\t");
    }).on('advance', function (e) {
      var time = e.data.time;

      if (time) {
        _this._setTime(time);
      }
    }).on('start:change end:change', function () {
      _this._slider.value = [_this._animation.startOffset() / (3600 * 1000), _this._animation.endOffset() / (3600 * 1000)];

      _this.trigger('change:range', {
        start: _this._animation.startDate(),
        end: _this._animation.endDate()
      });
    });

    if (this._indicator) {
      this._animation.on('load:start', function () {
        _this._indicator.startAnimating();
      }).on('load:done', function () {
        _this._indicator.stopAnimating();
      });
    }

    this._timeline.slider.on('change', function (e) {
      var progress = e.data.progress;

      if ((0, _utils.isset)(progress)) {
        var anim = _this._animation;
        var delta = anim.endDate().getTime() - anim.startDate().getTime();
        var time = Math.round(anim.startDate().getTime() + delta * progress);

        _this._setTime(time);
      }
    });

    this._slider.on('change:end', function (e) {
      var _a = e.data.value,
          min = _a[0],
          max = _a[1];

      _this._animation.setStartOffset(min * 3600 * 1000);

      _this._animation.setEndOffset(max * 3600 * 1000);
    });

    this.ui.play.on('click', function () {
      if (_this._animation.isAnimating() || _this._animation.isLoading()) {
        _this._animation.stop();
      } else {
        _this._animation.play();
      }
    });
    this.ui.now.on('click', function () {
      _this._animation.goToInit();
    });

    if (this.ui.settings) {
      this.ui.settings.on('click', function () {
        if (!_this._tooltip) {
          // const tooltip = new Tooltip(this._range.$el);
          var tooltip = new _Tooltip.default(_this._slider.$el);
          tooltip.attachTo(_this.ui.settings);
          _this._tooltip = tooltip;
        }

        if (_this._tooltip.visible) {
          _this._tooltip.hide();
        } else {
          _this._tooltip.show();
        }
      });
    }
  };

  TimelinePanel.prototype._setTime = function (time) {
    if (!this.ui.time) return;
    var date = new Date(time);
    this.ui.time.html("<p class=\"time\">" + (0, _dateFns.format)(date, 'h:mm a') + "</p><p class=\"day\">" + (0, _dateFns.format)(date, 'ddd, MMM D') + "</p>");
  };

  return TimelinePanel;
}(_Panel.default);

var _default = TimelinePanel;
exports.default = _default;
module.exports = exports.default;
