// @aerisweather/javascript-sdk/dist/maps/interactive/strategies/openlayers/layers/PolygonLayer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _Layer = _interopRequireDefault(require("./Layer"));

var strings = _interopRequireWildcard(require("../../../../../utils/strings"));

var _utils = require("../../../../../utils");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var __extends = void 0 && (void 0).__extends || function () {
  var extendStatics = function (d, b) {
    extendStatics = Object.setPrototypeOf || {
      __proto__: []
    } instanceof Array && function (d, b) {
      d.__proto__ = b;
    } || function (d, b) {
      for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    };

    return extendStatics(d, b);
  };

  return function (d, b) {
    extendStatics(d, b);

    function __() {
      this.constructor = d;
    }

    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
  };
}();
/* eslint-disable spaced-comment */
/// <reference path="../../../../../../node_modules/@types/openlayers/index.d.ts" />

/* eslint-enable spaced-comment */


var OpenLayersPolygonLayer =
/** @class */
function (_super) {
  __extends(OpenLayersPolygonLayer, _super);

  function OpenLayersPolygonLayer(layer, polygons, opts) {
    if (opts === void 0) {
      opts = undefined;
    }

    var _this = _super.call(this, layer, opts.id) || this;

    _this._opts = {};
    _this._polygons = polygons; // convert camelcase style keys to hypenated strings

    opts = opts || {};
    Object.keys(opts).forEach(function (key) {
      var k = strings.hyphenate(key);

      if (k !== key) {
        opts[k] = opts[key];
        delete opts[key];
      }
    });
    _this._opts = opts;
    return _this;
  }

  OpenLayersPolygonLayer.prototype.addTo = function (map, index) {
    _super.prototype.addTo.call(this, map, index);

    if (!this.layer) {
      this.invalidate();
    }
  };

  OpenLayersPolygonLayer.prototype.invalidate = function () {
    var id;
    var isMulti = this._polygons.length > 1;
    var polygons = (this._polygons || []).map(function (polygon) {
      if (!(0, _utils.isset)(id) && (0, _utils.isset)(polygon.id)) {
        id = polygon.id;
      }

      var outer = polygon.outer.map(function (c) {
        return [c.lon, c.lat];
      });
      var inner = polygon.inner ? polygon.inner.map(function (p) {
        return p.map(function (c) {
          return [c.lon, c.lat];
        });
      }) : [];
      return [outer].concat(inner);
    });
    this.source = new ol.source.Vector({
      features: new ol.format.GeoJSON().readFeatures({
        type: 'FeatureCollection',
        features: [{
          type: 'Feature',
          properties: {
            id: id
          },
          geometry: {
            type: isMulti ? 'MultiPolygon' : 'Polygon',
            coordinates: isMulti ? polygons : polygons[0] || []
          }
        }]
      }, {
        featureProjection: 'EPSG:3857'
      })
    });
    var _a = this._opts,
        stroke = _a.stroke,
        fill = _a.fill;
    this.layer = new ol.layer.Vector({
      source: this.source,
      style: new ol.style.Style({
        stroke: stroke ? new ol.style.Stroke(stroke) : null,
        fill: fill ? new ol.style.Fill(fill) : null
      })
    });

    if (this.map) {
      var map = this.map;
      this.remove();
      this.addTo(map, this._targetIndex);
    }
  };

  return OpenLayersPolygonLayer;
}(_Layer.default);

var _default = OpenLayersPolygonLayer;
exports.default = _default;
module.exports = exports.default;
