// @aerisweather/javascript-sdk/dist/network/api/ApiResult.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

/**
 * An `ApiResult` object contains response information about an API request.
 */
var ApiResult =
/** @class */
function () {
  /**
   * Initializes a result instance with the necessary response information.
   *
   * @param {*} response
   * @param {*} data
   * @param {*} error
   * @param {*} [params=null]
   * @memberof ApiResult
   */
  function ApiResult(response, data, error, params) {
    if (params === void 0) {
      params = null;
    }

    this.response = response;
    this.data = data;
    this.error = error;
    this.params = params;
  }
  /**
   * Returns the headers returned by the response, if any.
   *
   * @returns {*}
   * @memberof ApiResult
   */


  ApiResult.prototype.headers = function () {
    return this.response ? this.response.headers : undefined;
  };

  return ApiResult;
}();

var _default = ApiResult;
exports.default = _default;
module.exports = exports.default;
