// @aerisweather/javascript-sdk/dist/maps/geo/MultiLineString.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _LineString = _interopRequireDefault(require("./LineString"));

var _CoordinateBounds = _interopRequireDefault(require("../../geo/CoordinateBounds"));

var _utils = require("../../utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var MultiLineString =
/** @class */
function () {
  function MultiLineString(coords) {
    if ((0, _utils.isEmpty)(coords)) {
      throw new Error('Invalid coordinates for MultiLineString');
    }

    this.lines = coords.map(function (lineCoords) {
      return new _LineString.default(lineCoords);
    });
    this.coordinates = this.lines.map(function (line) {
      return line.coordinates;
    });
  }

  MultiLineString.prototype.toArray = function (reversed) {
    if (reversed === void 0) {
      reversed = false;
    }

    return this.lines.map(function (line) {
      return line.toArray(reversed);
    });
  };

  MultiLineString.prototype.bounds = function () {
    var bounds = {
      north: null,
      south: null,
      west: null,
      east: null
    };
    this.lines.forEach(function (line) {
      var b = line.bounds();

      if ((0, _utils.isset)(b)) {
        bounds.north = Math.max(bounds.north, b.north);
        bounds.south = Math.min(bounds.south, b.south);
        bounds.west = Math.min(bounds.west, b.west);
        bounds.east = Math.max(bounds.east, b.east);
      }
    });
    return _CoordinateBounds.default.fromBounds(bounds);
  };

  return MultiLineString;
}();

var _default = MultiLineString;
exports.default = _default;
module.exports = exports.default;
