// @aerisweather/javascript-sdk/dist/maps/views/legends/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LegendLabelPosition = exports.LegendOrientation = exports.LegendType = void 0;

/**
 * An enumerated value representing a legend output type.
 *
 * @export
 * @enum {string}
 */
var LegendType;
exports.LegendType = LegendType;

(function (LegendType) {
  LegendType["BAR"] = "bar";
  LegendType["POINT"] = "point";
})(LegendType || (exports.LegendType = LegendType = {}));
/**
 * An enumerated value representing a legend layout orientation.
 *
 * @export
 * @enum {string}
 */


var LegendOrientation;
exports.LegendOrientation = LegendOrientation;

(function (LegendOrientation) {
  /** Vertical layout. */
  LegendOrientation["VERTICAL"] = "v";
  /** Horizontal layout. */

  LegendOrientation["HORIZONTAL"] = "h";
})(LegendOrientation || (exports.LegendOrientation = LegendOrientation = {}));
/**
 * An enumerated value representing a legend label position.
 *
 * @export
 * @enum {string}
 */


var LegendLabelPosition;
exports.LegendLabelPosition = LegendLabelPosition;

(function (LegendLabelPosition) {
  /** Labels positioned below a legend color bar. Only supported with a horizontal legend orientation. */
  LegendLabelPosition["TOP"] = "t";
  /** Labels positioned above the legend color bar. Only supported with a horizontal legend orientation. */

  LegendLabelPosition["BOTTOM"] = "b";
  /** Labels positioned to the left of the legend color bar. Only supported with a vertical legend orientation. */

  LegendLabelPosition["LEFT"] = "l";
  /** Labels positioned to the right of the legend color bar. Only supported with a vertical legend orientation. */

  LegendLabelPosition["RIGHT"] = "r";
})(LegendLabelPosition || (exports.LegendLabelPosition = LegendLabelPosition = {}));
