// @aerisweather/javascript-sdk/dist/maps/interactive/strategies/google/sources/GoogleTileSource.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _TileSource = _interopRequireDefault(require("../../../sources/TileSource"));

var _utils = require("../../../../../utils");

var _utils2 = require("../../../../utils");

var _TileLayer = _interopRequireDefault(require("../layers/TileLayer"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var __extends = void 0 && (void 0).__extends || function () {
  var extendStatics = function (d, b) {
    extendStatics = Object.setPrototypeOf || {
      __proto__: []
    } instanceof Array && function (d, b) {
      d.__proto__ = b;
    } || function (d, b) {
      for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    };

    return extendStatics(d, b);
  };

  return function (d, b) {
    extendStatics(d, b);

    function __() {
      this.constructor = d;
    }

    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
  };
}();
/* eslint-disable spaced-comment */
///<reference path="../../../../../../node_modules/@types/googlemaps/index.d.ts" />

/* eslint-enable spaced-comment */


/**
 * `GoogleTileSource` is a subclass of `TileSource` that creates and manages a single tile layer to
 * be used with an `GoogleMapStrategy`.
 *
 * @export
 * @class LeafletTileSource
 * @extends {TileSource}
 */
var GoogleTileSource =
/** @class */
function (_super) {
  __extends(GoogleTileSource, _super);

  function GoogleTileSource(key, service, opts) {
    var _this = _super.call(this, key, service, opts) || this;

    service.offset(0);
    _this._overlay = new _TileLayer.default(key);
    _this.timestamp = _this._opts.time;
    _this.animation.provider = _this; // need to listen for reset events on the animation so that we can remove any actively
    // loading Google layer while in the middle of loading

    _this.animation.on('reset', function () {
      if (_this._loadingSource) {
        _this.trigger('layer:remove', {
          layer: _this._loadingSource.overlay
        });
      }
    });

    return _this;
  }

  Object.defineProperty(GoogleTileSource.prototype, "overlay", {
    get: function () {
      return this._overlay;
    },
    enumerable: true,
    configurable: true
  });
  Object.defineProperty(GoogleTileSource.prototype, "timestamp", {
    set: function (value) {
      value = new Date(value.getFullYear(), value.getMonth(), value.getDate(), value.getHours(), value.getMinutes()); // eslint-disable-line max-len

      if (!this._timestamp || value.getTime() !== this._timestamp.getTime()) {
        this._timestamp = value;

        if (this.isAnimating() === false) {
          this.updateVisibility();
        }

        if (this.overlay && this.overlay.isHidden === false && this.canShow()) {
          this.overlay.setUrl(this.urlTemplate());
        }
      }
    },
    enumerable: true,
    configurable: true
  });
  /**
   * Shows the layer associated with the source.
   *
   * @memberof GoogleTileSource
   */

  GoogleTileSource.prototype.show = function () {
    this._hidden = false;

    if (this.overlay) {
      this.overlay.setOpacity(this._opacity);
    }
  };
  /**
   * Hides the layer associated with the source.
   *
   * @memberof GoogleTileSource
   */


  GoogleTileSource.prototype.hide = function () {
    this._hidden = true;

    if (this.overlay) {
      this.overlay.setOpacity(0);
    }
  };
  /**
   * Flags the source for removal from the managing map strategy.
   *
   * @memberof GoogleTileSource
   */


  GoogleTileSource.prototype.remove = function () {
    this.trigger('layer:remove', {
      layer: this.overlay
    });
  };
  /**
   * Changes the opacity of source's layer.
   *
   * @param {number} value
   * @memberof GoogleTileSource
   */


  GoogleTileSource.prototype.setOpacity = function (value) {
    this._opacity = value;

    if (this.overlay) {
      this.overlay.setOpacity(value);
    }
  };
  /**
   * Changes the z-index of the source's layer.
   *
   * @param {number} value
   * @memberof GoogleTileSource
   */


  GoogleTileSource.prototype.setOrder = function (value) {
    console.warn('[Aeris] - Google Maps does not currently support changing the zIndex for an existing tile layer');

    if (this.node && this.node instanceof HTMLElement) {
      this.node.style.zIndex = value.toString();
    }
  };
  /**
   * Animation Providers
   */


  GoogleTileSource.prototype.animationLayerForDate = function (animation, date) {
    var _this = this;

    return new Promise(function (resolve, reject) {
      if (_this.dataSource) {
        var source_1 = _this.dataSource.tileLayer(_this.key, date, (0, _utils.extend)({}, _this._opts, {
          id: _this.identifier + "-anim-" + (0, _utils2.mapTimestampFromDate)(date)
        }));

        source_1.setOpacity(_this._opacity);
        source_1.hide();
        _this._loadingSource = source_1;
        source_1.overlay.on('load', function () {
          _this._loadingSource = null;
          resolve(source_1);
        }); // forward source layer remove event up to the parent source

        source_1.on('layer:remove', function () {
          _this.trigger('layer:remove', {
            layer: source_1.overlay
          });
        });

        _this.trigger('layer:add', {
          layer: source_1.overlay
        });
      } else {
        reject(new Error('No map strategy data source provided'));
      }
    });
  };

  return GoogleTileSource;
}(_TileSource.default);

var _default = GoogleTileSource;
exports.default = _default;
module.exports = exports.default;
